Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcContentServer.cs

```diff
@@ -344,6 +360,94 @@ private async Task<RequestCopyFileResponse> RequestCopyFileAsync(RequestCopyFile
             return new RequestCopyFileResponse { Header = ResponseHeader.Failure(startTime, $"No stores implement {nameof(ICopyRequestHandler)}.") };
         }
 
+        private readonly ConcurrentBigSet<ContentHash> _ongoingPushes = new ConcurrentBigSet<ContentHash>();
+
+        /// <summary>
+        /// Handles a request to copy content to this machine.
+        /// </summary>
+        public async Task HandlePushFileAsync(IAsyncStreamReader<PushFileRequest> requestStream, IServerStreamWriter<PushFileResponse> responseStream, ServerCallContext callContext)
+        {
+            OperationStarted();
+
+            var startTime = DateTime.UtcNow;
+
+            byte[] hashBytes = null;
+            var hashType = HashType.Unknown;
+            Guid traceId = default;
+            foreach (var header in callContext.RequestHeaders)

```

Comment:
Rather than using headers directly in server and client, I recommend creating a type (e.g. PushRequest) that stores then, and knows how to construct itself from headers via a static FromHeaders method and how to convert itself to headers via an instance ToHeaders method. Then use the type in both client and server. This keeps all the serialization/deserialization logic (e.g. header names) in one place.